### DISCUSSION TITLE
Simplifying the Public API (`HamtForest`)

### DISCUSSION THREAD

**Discussion by DougAnderson444** [wnfs-wg/rs-wnfs#407] [Ideas]
I'm digging into WNFS a bit, and what I'm noticing is there are a lot of options:

```
HamtForest:
new_rc
new_rsa_2048
new_rsa_2048_rc
new_trusted
new_trusted_rc
```

Any chance these could be slimmed down to make getting up and running faster with fewer choices? Honestly I find it hard to choose which one to use.

**Comment by matheus23** [comment]
Yeah, APIs could use some work. At the moment we're in "make it work" phase, and not yet in "make it nice".

I consider almost all of the API to be low-level at the moment, requiring you to manage the mutable root pointer to the latest state of the file system itself, and requiring you to keep around pointers to the BlockStore, forest and randomness, too.

The plan is to expose the "RootTree" API eventually. At the moment it's a hidden module (due to being unfinished and missing documentation). You can see the latest efforts in this PR: https://github.com/wnfs-wg/rs-wnfs/pull/390
To get it over the line it needs:
- a standardized root block serialization format that's compatible with wnfs-wg/nest
- an improved shared data API
- transactions & concurrency handling, but that needs conflict reconciliation to be implemented
- a bunch of documentation
